modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/SqlRowHandler.java

```diff
@@ -0,0 +1,272 @@
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public RowWrapper create(ByteBuffer buf) {
+                return wrap(new BinaryTuple(rowLen, buf));
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public RowWrapper wrap(InternalTuple tuple) {
+                assert rowLen == tuple.elementCount();
+
+                return new BinaryTupleRowWrapper(rowSchema, tuple);
+            }
+        };
+    }
+
+    /**
+     * Provides the ability for a single {@link RowHandler} instance to interact with multiple row implementations.
+     */
+    public abstract static class RowWrapper {
+        abstract int columnsCount();
+
+        abstract RowSchema rowSchema();
+
+        abstract @Nullable Object get(int field);
+
+        abstract void set(int field, Object value);
+
+        abstract ByteBuffer toByteBuffer();
+    }
+
+    /**
+     * Wrapper over an array of objects.
+     */
+    private static class ObjectsArrayRowWrapper extends RowWrapper {
+        private final RowSchema rowSchema;
+        private final Object[] row;
+
+        ObjectsArrayRowWrapper(RowSchema rowSchema, Object[] row) {
+            this.rowSchema = rowSchema;
+            this.row = row;
+        }
+
+        @Override
+        int columnsCount() {
+            return row.length;
+        }
+
+        @Override
+        @Nullable Object get(int field) {
+            return row[field];
+        }
+
+        @Override
+        void set(int field, @Nullable Object value) {
+            row[field] = value;
+        }
+
+        @Override
+        ByteBuffer toByteBuffer() {
+            return RowConverter.toByteBuffer(rowSchema, row);
+        }
+
+        @Override
+        RowSchema rowSchema() {
+            return rowSchema;
+        }
+    }
+
+    /**
+     * Wrapper over an {@link BinaryTuple}.
+     *
+     * <p>Since {@link BinaryTuple binary tuple} is immutable this wrapper doesn't support {@link #set(int, Object)} operation.
+     */
+    private static class BinaryTupleRowWrapper extends RowWrapper {
+        private final RowSchema rowSchema;
+        private final InternalTuple tuple;
+
+        BinaryTupleRowWrapper(RowSchema rowSchema, InternalTuple tuple) {
+            this.rowSchema = rowSchema;
+            this.tuple = tuple;
+        }
+
+        @Override
+        int columnsCount() {
+            return tuple.elementCount();
+        }
+
+        @Override
+        @Nullable Object get(int field) {
+            Object value = rowSchema.value(tuple, field);
+
+            if (value == null) {
+                return null;
+            }
+
+            NativeType nativeType = RowSchemaTypes.toNativeType(rowSchema.fields().get(field));
+
+            return TypeUtils.toInternal(value, Commons.nativeTypeToClass(nativeType));
+        }
+
+        @Override
+        void set(int field, @Nullable Object value) {
+            throw new UnsupportedOperationException();
```

Comment:
how to make sure no one will invoke `set` for row that is wrapper around BinaryTuple?

Reply:
I agree that this method should be implemented or should not be provided at all.
From my point of view, the only right way is to replace this method with append(Object obj) and rework all related code.